Exercicio 

A Terra está superlotada e a agência espacial "JavaSpace" está vendendo lotes em novos planetas. O preço de um lote é baseado em um "Índice de Habitabilidade" (0 a 100).

Sua missão é criar um modelo de Regressão Linear Simples em Java para prever o Índice de Habitabilidade de um planeta com base na sua Distância da Estrela Solar (em Unidades Astronômicas).


Utilizem os seguintes dados históricos de planetas já colonizados:

 
-----Planeta-----                                 

Alpha-1                                                            

Beta-2                                                              

Gamma-3                                                        

Delta-4                                                        

Epsilon-5                


-----Distância (x - em UA)------    

 0.8                           

 1.2                             

 1.5                           
 
 0.5  

 2.0                             


 
-----Índice de Habitabilidade (y) -----

85

70

60

40 (quente demais!)

30 (frio demais!)

================================================ 

-- O Desafio Prático


Tarefa 1: O Algoritmo "Raça Pura"



Não use bibliotecas externas (como Weka ou DeepNetts) num primeiro momento. Implemente uma classe RegressaoEspacial que:



Receba dois arrays de double (X e Y).

Calcule a média de X e Y.

Calcule os coeficientes β0 e β1.

Tenha um método prever(double distancia) que retorne o índice.

In [2]:
import numpy as np

In [ ]:
planet = np.array(['Alpha-1', 'Beta-1', 'Gamma-3', 'Delta-4', 'Epsilon-5'])

distance_x = np.array([0.8, 1.2, 1.5, 0.5, 2.0])

indice_hab_y = np.array([85,  70,  60, 40, 30])

In [10]:
#CALCULANDO A MÉDIA DE X E Y
mean_x = np.mean(distance_x)

mean_y = np.mean(indice_hab_y)

In [15]:
#Média
print(f"Média distância x: {mean_x} ")

#Média Y
print(f"Média índice de habitalidade: {mean_y}")

Média distância x: 1.2 
Média índice de habitalidade: 57.0


In [11]:
#Calculando os coeficientes
b1 = np.sum((distance_x - mean_x) * (indice_hab_y - mean_y)) / np.sum((distance_x - mean_x) ** 2)

b0 = mean_y - b1 * mean_x

In [13]:
#Resultado inclinação
print(f"Resultado inclinação: {b1}")

#Resultado Intercepto
print(f"Intercepto: {b0}")

Resultado inclinação: -14.492753623188406
Intercepto: 74.3913043478261


In [17]:
#Calculando método prever
def foresees(distance_x_, b1, b0):
    return b0 + (b1 * distance_x_)

In [34]:
#Previsão
forecast =  foresees(1.1, b0, b1)
print(f"Previsão Z-java : {forecast:.2f}")

Previsão Z-java : 67.34


In [38]:
#Previsão para todos os planetas
calc_forecast_all = b0 + (b1 * distance_x)

In [39]:
print(calc_forecast_all)

[62.79710145 57.         52.65217391 67.14492754 45.4057971 ]


In [43]:
mse = np.mean((indice_hab_y - calc_forecast_all) ** 2)

In [44]:
print(f"MSE: {mse:.2f}")

MSE: 338.03


*a) O modelo é perfeito? Por quê?*
       
R: Um modelo MSE de 338.03 mostra que tem margem de erro. Isso acontece porque: Temos 5 planetas de dados, muito pouco!

*b) Como o Java se comportaria com 1 bilhão de planetas?*
        
R: Iria se comportar bem, isso acontece por que teremos mais planetas, e com mais planetas temos mais dados. Outra questão do comportamento será o valor do MSE, pois ele irá cair drasticamente. Com mais dados, o modelo consegue aprender o padrão real.



*c) Quanto a outras variáveis, que outras colunas poderíamos adicionar?*

    - Temperatura média do planeta; 

    - Presença de agua; 
    
    - Porcentagem oxigênio
    
    - Gravidade; 
    
    - Tamanho do planeta; 
    
    - Terra Fértil;